sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst

```diff
@@ -0,0 +1,162 @@
+SYCL program. The library should match the C or C++ standard library
+used to compile the program:
+
+For example, on Linux with GNU glibc:
+.. code:
+   clang++ -fsycl -c main.cpp -o main.o
+   clang++ -fsycl main.o $(SYCL_INSTALL)/lib/libsycl-glibc.o -o a.out
+
+or, in case of Windows:
+.. code:
+   clang++ -fsycl -c main.cpp -o main.obj
+   clang++ -fsycl main.obj %SYCL_INSTALL%/lib/libsycl-msvc.o -o a.exe
+
+List of supported functions from C standard library:
+  - assert macro (from <assert.h> or <cassert>)
+
+NOTE: only the GNU glibc and Microsoft C libraries are currently
+supported.
+
+Example of usage
+================
+
+.. code: c++
+   #include <assert.h>
+   #include <CL/sycl.hpp>
+
+   template <typename T, size_t N>
+   void simple_vadd(const std::array<T, N>& VA, const std::array<T, N>& VB,
+                    std::array<T, N>& VC) {
+     // ...
+     cl::sycl::range<1> numOfItems{N};
+     cl::sycl::buffer<T, 1> bufferA(VA.data(), numOfItems);
+     cl::sycl::buffer<T, 1> bufferB(VB.data(), numOfItems);
+     cl::sycl::buffer<T, 1> bufferC(VC.data(), numOfItems);
+
+     deviceQueue.submit([&](cl::sycl::handler& cgh) {
+       auto accessorA = bufferA.template get_access<sycl_read>(cgh);
+       auto accessorB = bufferB.template get_access<sycl_read>(cgh);
+       auto accessorC = bufferC.template get_access<sycl_write>(cgh);
+
+       cgh.parallel_for<class SimpleVadd<T>>(numOfItems,
+       [=](cl::sycl::id<1> wiID) {
+           accessorC[wiID] = accessorA[wiID] + accessorB[wiID];
+           assert(accessorC[wiID] > 0 && "Invalid value");
+       });
+     });
+     deviceQueue.wait_and_throw();
+   }
+
+Frontend
+========
+
+Once the system header is included, the corresponding functions can be
+used in SYCL device code. This results in a handful of unresolved
+functions in LLVM IR after clang:
+
+.. code:
+    ; Function Attrs: noreturn nounwind
+    declare dso_local spir_func void @__assert_fail(i8 addrspace(4)*, i8 addrspace(4)*, i32, i8 addrspace(4)*)
+
+    [...]
+    cond.false:
+      call spir_func void @__assert_fail([...])
+      unreachable
+
+The C and C++ specifications do not define names and signatures of the
+functions from libc implementation that are used for a particular
+function. For example, the `assert` macro:
+
+  - in Glibc and musl libraries it expands to `__assert_fail`
+  - in MSVC library it expands to `_wassert`
+  - in newlib library it expands to `__assert_func`
+
+This makes it difficult to handle all possible cases in device
+compilers. In order to facilitate porting to new platforms, and to
+avoid imposing a lot of boilerplate code in *every* device compiler,
+wrapper libraries are provided with the SYCL compiler that "lower"
+libc implementation-specific functions into a stable set of functions,
+that can be later handled by a device compiler.
+
+.. code:
+   clang++ -fsycl -c main.cpp -o main.o
+   clang++ -fsycl main.o $(SYCL_INSTALL)/lib/libsycl-glibc.o -o a.out
```

Comment:
Two points:
1. Do we really need a separate wrapper for each standard library implementation? I don't think I understand the problem with a `boilerplate code in *every* device compiler`. I think linker is able to import only necessary functions, so scattering "boilerplate" code into separate wrapper libraries seems unnecessary.
2. If we use single device library, it can be linked by the driver implicitly, but even if we stick to existing proposal, driver is able to infer right library from the host triple.

Reply:
(1) should be explained on line 77. Let me know if I should elaborate.
(2) can be done as a separate change in driver.

Reply:
(1) I understand that we should have definitions for `__assert_fail`, `_wassert` and `__assert_func` to support Glibc, MSVC and newlib, but can we keep them all in the same library? What is the strategy for the symbols shared between different implementation (if any): e.g. Glibc and newlib?

Reply:
> (1) I understand that we should have definitions for `__assert_fail`, `_wassert` and `__assert_func` to support Glibc, MSVC and newlib, but can we keep them all in the same library? 

This doesn't work for functions that have the same name, but different arguments (e.g. __assert_fail in glibc vs musl).

> What is the strategy for the symbols shared between different implementation (if any): e.g. Glibc and newlib?

There are no shared symbols. If there is a function with identical name and signature, it will be duplicated in every wrapper library.

In any case, I anticipate the wrapper libraries to go away soon. @s-kanaev is working on a header-based approach to provide the wrappers.